Automatically generate a full coupling matrix for global AD indexing

lindsayad opened the pull request:
This was kind of a beast. Per the description in #16396 we have to make
sure we trigger the update to a full coupling matrix before the problem
is init'd in order to make sure that we don't have to reinit later with
a different coupling matrix (EquationSystems init and reinit are expensive).
Consequently, I leverage the `RelationshipManager` system.

Closes #16396

lindsayad committed:
Automatically generate a full coupling matrix for global AD indexing

This was kind of a beast. Per the description in #16396 we have to make
sure we trigger the update to a full coupling matrix before the problem
is init'd in order to make sure that we don't have to reinit later with
a different coupling matrix (EquationSystems init and reinit are expensive).
Consequently, I leverage the `RelationshipManager` system.

Closes #16396

lindsayad committed:
Add tests (#16396)

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 4f3fa5a6a2104ab8784a6519677589738b9aef6f
+Subproject commit ffc1d2d72bc35a75b798602ba3e195e5481e9f8f
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/641228/) on 9353f97 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16397/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Pthreads+OpenMP Test](https://civet.inl.gov/job/644185/) on 01073d8 : invalidated by @lindsayad

Another why

fdkong commented on framework/include/relationshipmanagers/ElementSideNeighborLayers.h:
```diff
@@ -42,7 +42,7 @@ class ElementSideNeighborLayers : public FunctorRelationshipManager
 
   virtual std::string getInfo() const override;
-  virtual bool operator==(const RelationshipManager & rhs) const override;
+  virtual bool operator>=(const RelationshipManager & rhs) const override;
```
This is a great change to reduce the number of ghost functors

rwcarlsen commented on framework/src/interfacekernels/InterfaceKernelBase.C:
```diff
@@ -63,10 +64,14 @@ InterfaceKernelBase::validParams()
+      "ElementSideNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+          Moose::RelationshipManagerType::COUPLING,
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
+        rm_params.set<bool>("create_full_coupling_matrix") =
+            obj_params.get<bool>("is_ad") && Moose::globalADIndexing();
+      });
```
Could we default to doing this with global indexing and using AD?  And then provide a simple way for the user to override/disable the full coupling if they don't want it?

lindsayad replied:
This PR does already provide the means to override full coupling. I think you're right that we can just default to full coupling when `_fe_problem.haveADObjects()`. It's an excellent question how I settled on this method when what you proposed is soooo much simpler and does the exact same thing. A rabbit hole of relationship manager thinking I guess :laughing: Anyway yea I'll remove all of this logic. And then there will still be some good RM changes that come out of this PR anyways

lindsayad replied:
See https://github.com/idaholab/moose/pull/16397#issuecomment-747764789

lindsayad replied:
This kind of thing should only be done at the framework level in core residual/Jacobian computing classes. Application developers cannot add new residual/Jacobian objects that do not inherit from the core framework residual/Jacobian classes. The only time someone would have to reproduce this syntax that I've added is if they're adding some new core residual/Jacobian computing system. 

Because the problem initialization happens before we construct our residual objects, our solution essentially *has* to come out of `(Derived)ResidualObject::validParams`, unless you have a more clever idea. I do like the idea of toggling a change in variable coupling through (coupling) `RelationshipManagers/GhostingFunctors` since this is conceptually what they are supposed to handle: controlling construction of the sparsity pattern.

fdkong reviewed (dismissed):
Looks good

lindsayad commented:
I remembered why I did the full matrix coupling triggering through RelationshipManagers. Its because the sparsity pattern is initialized  well before we actually construct kernels etc (and hence well before we could use those constructors to tell the problem we have AD objects). And I guess because we determine the inter-element (and intra-element) coupling via relationship managers, it felt natural to set inter-variable coupling through relationship managers (in this case setting full variable coupling)

lindsayad committed:
Remove RelationshipManager triggering of full coupling for global AD

I don't think this is correct. We need to trigger full coupling before
`FEProblemBase::init` and it feels natural to me for coupling functors
to be the ones that modify variable coupling.

EDIT: This works fine now because we have moved the `init_problem` task
after addition of residual objects

lindsayad committed:
Init problem after addition of residual objects

Now we can simply query whether we have any AD objects to determine
whether we should bump our coupling matrix to a full coupling matrix

lindsayad commented:
Ok @rwcarlsen I've moved the `init_problem` task after adding of residual objects (and other objects), so now I can toggle creation of the full coupling matrix just based off the presence of AD objects. Nice idea!

lindsayad commented on framework/include/base/MooseApp.h:
```diff
@@ -708,6 +708,11 @@ class MooseApp : public ConsoleStreamInterface, public libMesh::ParallelObject
   bool addRelationshipManager(std::shared_ptr<RelationshipManager> relationship_manager);
 
+  /**
+   * Purge this relationship manager from meshes and DofMaps and finally from us
```
Would you prefer `us -> this` ?

lindsayad commented on framework/include/relationshipmanagers/RelationshipManager.h:
```diff
@@ -87,9 +87,29 @@ class RelationshipManager : public MooseObject, public libMesh::GhostingFunctor
+    mooseDeprecated(
+        "Are you overriding RelationshipManager::operator==? We are transiting to make operator>= "
+        "the required override. If you are not overriding operator>= or operator==, you are about "
+        "to get an error message saying that operator>= must be overridden.");
```
In the code it is `operator>=` without a space, so I would prefer to keep this as-is

lindsayad committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/652407/) on 2e9694c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16397/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16397/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8016b897f123725f2c49c83869cde7daad482e3`

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/652446/) on cfca91d : invalidated by @lindsayad

moosebuild commented:
Job [Modules debug PETSc alt](https://civet.inl.gov/job/652437/) on cfca91d : invalidated by @lindsayad

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/652446/) on cfca91d : invalidated by @lindsayad

rwcarlsen commented:
Those are definitely persistent failures :-(  Too bad it's Griffin.

lindsayad commented:
Yea I have Griffin patches merged into devel. Waiting for them to merge into master, and then I should be able to invalidate and get passing here.

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/652441/) on cfca91d : invalidated by @lindsayad

moosebuild commented:
Job [Controlled App tests PETSc alt](https://civet.inl.gov/job/652444/) on cfca91d : invalidated by @lindsayad
